### PR TITLE
Improve intrusive rendering

### PR DESCRIPTION
**Pull request by jan-br**

### CODE REVIEW

File: render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java [hunk 1]
```diff
@@ -42,6 +45,8 @@
    */
   boolean isIntrusive();
 
+  Collection<ScreenName> getIntrusiveScreens();
```
Comment (Janrupf) [inline render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java:54]: Missing documentation

File: render/webgui/src/internal/java/net/flintmc/render/gui/webgui/internal/MinecraftWebFileSystem.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+@Singleton
+@WebFileSystem("minecraft")
+public class MinecraftWebFileSystem implements WebFileSystemHandler {
+
+  private final File savesDirectory;
+
+  @Inject
+  private MinecraftWebFileSystem() {
+    this.savesDirectory = new File(".");
```
Comment (Janrupf) [inline render/webgui/src/internal/java/net/flintmc/render/gui/webgui/internal/MinecraftWebFileSystem.java]: Despite this entire class being an unrelated change, this should not use `.`, but rather the `--gameDir` argument passed to minecraft and only fall back to `.` in case the argument could not be found.

File: render/webgui/src/internal/java/net/flintmc/render/gui/webgui/internal/MinecraftWebFileSystem.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  public boolean existsFile(String path) {
+    return new File(savesDirectory, path).exists();
+  }
+
+  @Override
+  public WebResource getFile(String path) throws FileNotFoundException {
+    try {
+      return new URLWebResource(path, new File(savesDirectory, path).toURI().toURL());
+    } catch (MalformedURLException e) {
+      throw new RuntimeException(e);
```
Comment (Janrupf) [inline render/webgui/src/internal/java/net/flintmc/render/gui/webgui/internal/MinecraftWebFileSystem.java]: Throw a `FileNotFoundException`

File: render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java [hunk 4]
```diff
@@ -42,6 +45,8 @@
    */
   boolean isIntrusive();
```
Comment (Janrupf) [inline render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java:45]: From what I can tell, this can be inferred based on whether `getIntrusiveScreens()` is empty or not

File: render/gui/src/internal/java/net/flintmc/render/gui/internal/windowing/InternalWindow.java [hunk 5]
```diff
@@ -34,7 +34,7 @@
    *
    * @return {@code true} if the windows is renderer intrusively, {@code false} otherwise
    */
-  boolean isRenderedIntrusively();
+  boolean isRenderedIntrusively(String screen);
```
Comment (Janrupf) [inline render/gui/src/internal/java/net/flintmc/render/gui/internal/windowing/InternalWindow.java]: This should probably accept a `ScreenName` instead of a String

**Review by Janrupf — CHANGES_REQUESTED** [review]

File: render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java (hunk 4 above)
Comment (jan-br) [reply render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java:45]: Will be fixed, if false is returned, isIntrusive is meant to disable intrusive rendering completely even if getIntrusiveScreens() does contain the rendered screen.

File: render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java [hunk 6]
```diff
@@ -19,38 +19,48 @@
...
+     * Called when window needs to be rendered. The OpenGL context in this method is the one used for
+     * the window.
+     */
+    void render();
+
+    /**
+     * Called when the renderer is removed from a window. The OpenGL context in this method is the one
+     * used for the window.
+     */
+    void cleanup();
```
Comment (Janrupf) [inline render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java]: Looks like you used 4 spaces here

File: render/webgui/src/internal/java/net/flintmc/render/gui/webgui/internal/MinecraftWebFileSystem.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+@Singleton
+@WebFileSystem("minecraft")
+public class MinecraftWebFileSystem implements WebFileSystemHandler {
+
+  private final File gameDirectory;
+
+  @Inject
+  private MinecraftWebFileSystem(@Named("launchArguments") Map launchArguments) {
+    this.gameDirectory = new File(launchArguments.get("--gamedir").toString());
```
Comment (Janrupf) [inline render/webgui/src/internal/java/net/flintmc/render/gui/webgui/internal/MinecraftWebFileSystem.java]: **Fall back** to `.` in case this does not exist, as discussed in the previous conversation

File: render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java [hunk 8]
```diff
@@ -19,38 +19,48 @@
...
+    void initialize();
+
+    /**
+     * Determines if this renderer is taking full control of the content of the window.
+     * If this setting is {@code true} the rendering of every minecraft window in {@link WindowRenderer#getIntrusiveScreens()}
+     * will be cancelled when this renderer is registered.
+     *
+     * @return {@code true} to mark the renderer as intrusive, {@code false} to mark it as cooperative
+     */
+    boolean isIntrusive();
```
Comment (Janrupf) [inline render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java]: As discussed before, is this method really required? This should (from my understanding) only be `true`, if `getIntrusiveScreens().size() > 0`. To avoid bad implementations, this method should be removed and checking should always be done with aforementioned check

**Review by Janrupf — CHANGES_REQUESTED** [review]

File: render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java (hunk 8 above)
Comment (Janrupf) [reply render/gui/src/main/java/net/flintmc/render/gui/windowing/WindowRenderer.java]: NOTE: After clarification, it was decided to keep the method and make it periodically polled to check if results from `getIntrusiveScreens()`should be used

**Review by Janrupf — APPROVED** [review]
